tfjs-layers/src/layers/nlp/tokenizers_utils.ts

```diff
@@ -0,0 +1,139 @@
+
+  const bs = [
+    ...range('!'.charCodeAt(0), '~'.charCodeAt(0)),
+    ...range('¡'.charCodeAt(0), '¬'.charCodeAt(0)),
+    ...range('®'.charCodeAt(0), 'ÿ'.charCodeAt(0))
+  ];
+
+  const cs = [...bs];
+  let n = 0;
+
+  // Removes mapping an int to a whitespace character
+  for (let b = 0; b < 2 ** 8; b++) {
+    if (!bs.includes(b)) {
+      bs.push(b);
+      cs.push(2 ** 8 + n);
+      n++;
+    }
+  }
+
+  const chars = cs.map(n => String.fromCharCode(n));
+
+  // TODO(orderique): Verify same functionality.
+  const bytes = Uint8Array.from(bs);
+
+  return [bytes, chars];
+}
+
+/**
+ * StaticHashTable extends Map and includes a `lookup` function that
+ */
+class StaticHashTable<K, V extends number|string> {
+  private _map: Map<K, V>;
+
+  constructor(keys: K[], values: V[], private readonly defaultValue: V) {
+    if (keys.length !== values.length) {
+      throw new ValueError(`keys and values arrays must be same length.
+        Instead got lengths ${keys.length} and ${values.length}.`
+      );
+    }
+    const keyValPairs: Array<[K, V]> = [];
+    for (let idx = 0; idx < keys.length; idx++) {
+      const key = keys[idx];
+      const val = values[idx];
+      keyValPairs.push([key, val]);
+    }
+
+    this._map = new Map(keyValPairs);
+  }
+
+  get(key: K): V {
+    return this._map.get(key) || this.defaultValue;
+  }
+
+  lookup(keys: Tensor[]): Tensor[] {
+    const values = keys.map(tensor => {
+      const innerValues: V[] = [];
+      for (const key of tensor.dataSync() as unknown as K[]) {
+        innerValues.push(this.get(key));
+      }
+
+      return tensor1d(innerValues as string[]|number[]);
+    });
+
+    return values;
+  }
+}
+
+export function createStaticHashtable<T1, T2 extends number|string>(
+  keys: T1[], values: T2[], defaultVal: T2): StaticHashTable<T1, T2> {
+
+  return new StaticHashTable(keys, values, defaultVal);
+}
+
+/**
+ * Cache that stores the encoded result of seen tokens.
+ *
+ * The cache key is string tensor or python strings, and the value is split
+ * tokens joined by whitespace. For example, "dragonfly" => "dragon fly"
+ *
+ * Examples:
+ *
+ * ```
+ * const cache = new BytePairTokenizerCache();
+ * cache.insert(["butterfly", "dragonfly"], ["but ter fly", "dragon fly"]);
+ * cache.lookup(["butterfly"]);
+ * ```
+ */
+export class BytePairTokenizerCache {
+  private _cache: Map<string, string>;
+
+  constructor() {
+    this._cache = new Map();
+  }
+
+  /**
+   * Insert token <=> encoded outputs pairs.
+   */
+  insert(keys: Tensor|string[], values: string[]): BytePairTokenizerCache {
+    const arrKeys = keys instanceof Tensor ?
+      keys.dataSync() as unknown as string[] : keys;
+
+    arrKeys.forEach((key, idx) => this._cache.set(key, values[idx]));
+    return this;
+  }
+
+  /**
+   * Look up the encoded outputs of given tokens.
+   */
+  lookup(keys: Tensor|string[]): string[] {
+    const arrKeys = keys instanceof Tensor ?
+      keys.dataSync() as unknown as string[] : keys;
```

Comment:
If lookup is used often, you may want to make this async and use `await data` instead of dataSync.